### PR TITLE
Big refactor to make better use of the STL

### PR DESCRIPTION
**Pull request by diogocp**
My original intention was just to add the album artist and mbid, but I found that heavy use of std::map could greatly simplify the code.

To-do: cache albumArtist, trackNumber and mbid.

### CODE REVIEW

**Comment by hrkfdn** [comment]
Hey, thanks a lot for your patches. I'm really torn here.. For one, the code as it is now is very bad. I started writing it when I barely knew any C++, which is why most of it could be written in C. It doesn't use many of the (complex) features C++ offers, so a refactoring is definitely in order.
On the other hand I am unsure which direction mpdas should head towards, because I could see myself migrating it to a C-only codebase, as well, which would probably be a lot easier as it is mostly C code at the moment.

Long story short: I will need some time to consider this. Either way, thanks a lot for your contribution. Should I decide to keep it a C++ I will most likely merge your changes, as most of them look pretty reasonable on first glance :)

**Comment by diogocp** [comment]
I understand. Unfortunately both libmpdclient and libcurl are written in C, which makes it easy to "pollute" the code with C-isms. I actually thought about just writing a new scrobbler from scratch in C++11, but it would probably be too much work for just a few extra features.

Anyway, I will keep working on this since I still need a few more features (at least the new cache format and Last.fm's pseudo-OAuth).
